pandas/_libs/tslibs/offsets.pyx

```diff
@@ -304,6 +304,15 @@ class _BaseOffset(object):
     _day_opt = None
     _attributes = frozenset(['n', 'normalize'])
 
+    def __init__(self, n=1, normalize=False):
+        n = self._validate_n(n)
+        object.__setattr__(self, "n", n)
+        object.__setattr__(self, "normalize", normalize)
+        object.__setattr__(self, "_cache", {})
+
+    def __setattr__(self, name, value):
+        raise AttributeError("DateOffset objects are immutable.")
```

Comment:
The standard approach is usually to use private attribute like `_n` and provide access via a properties. Is there a reason why that wouldn't make sense here?

Reply:
That would absolutely make sense, but since micro-benchmarks are a concern I prefer to avoid the `property` overhead.  The best-case would be the implementation in #18824, but I can't figure out the pickle errors there.

Reply:
OK, seems reasonable to me.